docs: rewrite quickstart.md (#884)
pingcap/parser: docs/quickstart.md

# Quickstart

This parser is highly compatible with MySQL syntax. You can use it as a library, parse a text SQL into an AST tree, and traverse the AST nodes.

In this example, you are going to build a project, which can extract all the column names from a text SQL.

## Prerequisites

- [Golang](https://golang.org/dl/) version 1.13 or above. You can follow the instructions in the official [installation page](https://golang.org/doc/install) (check it by `go version`)

## Create a Project

```bash

> $ mkdir colx && cd colx
> $ go mod init colx && touch main.go

Reviewer: ```suggestion
mkdir colx && cd colx
go mod init colx && touch main.go
```

```

## Import Dependencies

First of all, you need to use `go get` to fetch the dependencies:

```bash
$ go get -v github.com/pingcap/parser@v4.0.0-rc.1
```

> **NOTE**
>
> If you want to use advanced API on expression(which is a kind of AST node), such as numbers, string literals, booleans, nulls, etc., it is strongly recommended to use TiDB's `types` package with the following command:
>
> ```bash
> go get -v github.com/pingcap/tidb/types/parser_driver@v4.0.0-rc.1
> ```
> and import it in the directive of your golang source code:
> ```go
> import _ "github.com/pingcap/tidb/types/parser_driver"
> ```

Your directory should contains three files:
```
.
├── go.mod
├── go.sum
└── main.go
```

Now, open `main.go` with your favouriate editor, start coding!

## Parse SQL text

To convert a SQL text to an AST tree, you can:

> 1. use `parser.New()` to instantiate a parser and

Reviewer: consider linking to the API docs.

```suggestion
1. use [`parser.New()`](https://pkg.go.dev/github.com/pingcap/parser?tab=doc#New) function to instantiate a parser and
```

2. call the method `Parse(sql, charset, collation)` on it.

```go
package main

import (
	"fmt"
	"github.com/pingcap/parser"
	"github.com/pingcap/parser/ast"
	_ "github.com/pingcap/parser/test_driver"
)

func parse(sql string) (*ast.StmtNode, error) {
	p := parser.New()

	stmtNodes, _, err := p.Parse(sql, "", "")
	if err != nil {
		return nil, err
	}

	return &stmtNodes[0], nil
}

func main() {
	astNode, err := parse("SELECT a, b FROM t")
	if err != nil {
		fmt.Printf("parse error: %v\n", err.Error())
		return
	}
	fmt.Printf("%v\n", *astNode)
}
```

Test it with

```bash

> $ go run main.go

Reviewer: ```suggestion
go run main.go
```

```

You should get something like this:

```
&{{{{SELECT a, b FROM t}}} {[]} 0xc0000a1980 false 0xc00000e7a0 <nil> 0xc0000a19b0 <nil> <nil> [] <nil> <nil> none [] false false 0 <nil>}
```

> **NOTE**
>
> Here are a few things you might want to know:
> - To use a parser, a `parser_driver` is required. It decides how to parse the basic data types in SQL.
>

> >   You can use `github.com/pingcap/parser/test_driver` as the `parser_driver` for test. Again, if you need advanced features, please use the `test_driver` in TiDB(`go get -v github.com/pingcap/tidb/types/parser_driver@v4.0.0-rc.1` and import it).

Reviewer: ```suggestion
>   You can use `github.com/pingcap/parser/test_driver` as the `parser_driver` for test. Again, if you need advanced features, please use the `test_driver` in TiDB (`go get -v github.com/pingcap/tidb/types/parser_driver@v4.0.0-rc.1` and import it).
```

BTW why 4.0.0-rc.1 ?

Author: I found the newest [released parser version](https://github.com/pingcap/parser/releases) is 4.0.0-rc.1.

## Traverse AST Nodes

Now you get the AST tree root of a SQL statement. It is time to extract the column names with traverse.

Parser implements the interface `ast.Node` for each kind of AST node, i.e. SelectStmt, TableName, ColumnName. `ast.Node` provides a method `Accept(v Visitor) (node Node, ok bool)` to allow any struct that implemented `ast.Visitor` to traverse itself.

`ast.Visitor` is defined as follows:
```go
type Visitor interface {
	Enter(n Node) (node Node, skipChildren bool)
	Leave(n Node) (node Node, ok bool)
}
```

Now you can define your own visitor, `colX`(columnExtractor):

```go
type colX struct{
	colNames []string
}

func (v *colX) Enter(in ast.Node) (ast.Node, bool) {
	if name, ok := in.(*ast.ColumnName); ok {
		v.colNames = append(v.colNames, name.Name.O)
	}
	return in, false
}

func (v *colX) Leave(in ast.Node) (ast.Node, bool) {
	return in, true
}
```

Finally, wrap `colX` in a simple function:

```go
func extract(rootNode *ast.StmtNode) []string {
	v := &colX{}
	(*rootNode).Accept(v)
	return v.colNames
}
```

and slightly modifiy main function:

```go
func main() {
	if len(os.Args) != 2 {

> 		fmt.Printf("usage: colx '%%s' (%%s can be any SQL statement)\n")

Reviewer: ```suggestion
		fmt.Println("usage: colx 'SQL statement'")
```

		return
	}
	sql := os.Args[1]
	astNode, err := parse(sql)
	if err != nil {
		fmt.Printf("parse error: %v\n", err.Error())
		return
	}
	fmt.Printf("%v\n", extract(astNode))
}
```

Test it:

```bash
$ go build && ./colx 'select a, b from t'
```

```
[a b]
```

You can also try with different SQL as an input, i.e.

> ```bash
> $ ./colx 'SELECT a, b FROM t GROUP BY (a, b) HAVING a > c ORDER BY b'
> ```
> ```bash
> $ ./colx 'SELECT a, b FROM t/invalid_str'

Reviewer: ````suggestion
```console
$ ./colx 'SELECT a, b FROM t GROUP BY (a, b) HAVING a > c ORDER BY b'
[a b c]
$ ./colx 'SELECT a, b FROM t/invalid_str'
parse error: …
````

# Quickstart

This parser is highly compatible with MySQL syntax. You can use it as a library, parse a text SQL into an AST tree, and traverse the AST nodes.

In this example, you will build a project, which can extract all the column names from a text SQL.

## Prerequisites

- [Golang](https://golang.org/dl/) version 1.13 or above. You can follow the instructions in the official [installation page](https://golang.org/doc/install) (check it by `go version`)

## Create a Project

```bash
mkdir colx && cd colx
go mod init colx && touch main.go
```

## Import Dependencies

First of all, you need to use `go get` to fetch the dependencies:

```bash
go get -v github.com/pingcap/parser@v4.0.0-rc.1
```

> **NOTE**
>

> > If you want to use advanced API on expression(which is a kind of AST node), such as numbers, string literals, booleans, nulls, etc., it is strongly recommended to use TiDB's `types` package with the following command:

Reviewer 2: ```suggestion
> If you want to use advanced API on expression (which is a kind of AST node), such as numbers, string literals, booleans, nulls, etc. It is strongly recommended to use `types` package in TiDB repo with the following command:
```

To convert a SQL text to an AST tree, you can:
1. use [`parser.New()`](https://pkg.go.dev/github.com/pingcap/parser?tab=doc#New) function to instantiate a parser and
2. invoke the method [`Parse(sql, charset, collation)`](https://pkg.go.dev/github.com/pingcap/parser?tab=doc#Parser.Parse) on it.

```go
package main

import (
	"fmt"
	"github.com/pingcap/parser"
	"github.com/pingcap/parser/ast"
	_ "github.com/pingcap/parser/test_driver"
)

func parse(sql string) (*ast.StmtNode, error) {
	p := parser.New()

	stmtNodes, _, err := p.Parse(sql, "", "")
	if err != nil {
		return nil, err
	}

	return &stmtNodes[0], nil
}

func main() {
	astNode, err := parse("SELECT a, b FROM t")
	if err != nil {
		fmt.Printf("parse error: %v\n", err.Error())
		return
	}
	fmt.Printf("%v\n", *astNode)
}
```

Test it with

```bash
go run main.go
```

You should get something like this:

```
&{{{{SELECT a, b FROM t}}} {[]} 0xc0000a1980 false 0xc00000e7a0 <nil> 0xc0000a19b0 <nil> <nil> [] <nil> <nil> none [] false false 0 <nil>}
```

> **NOTE**
>
> Here are a few things you might want to know:
> - To use a parser, a `parser_driver` is required. It decides how to parse the basic data types in SQL.
>

> >   You can use [`github.com/pingcap/parser/test_driver`](https://pkg.go.dev/github.com/pingcap/parser/test_driver) as the `parser_driver` for test. Again, if you need advanced features, please use the `test_driver` in TiDB (`go get -v github.com/pingcap/tidb/types/parser_driver@v4.0.0-rc.1` and import it).

Reviewer 2: s/test_driver/parser-driver?

```go
type Visitor interface {
	Enter(n Node) (node Node, skipChildren bool)
	Leave(n Node) (node Node, ok bool)
}
```

Now you can define your own visitor, `colX`(columnExtractor):

```go
type colX struct{
	colNames []string
}

func (v *colX) Enter(in ast.Node) (ast.Node, bool) {
	if name, ok := in.(*ast.ColumnName); ok {
		v.colNames = append(v.colNames, name.Name.O)
	}
	return in, false
}

func (v *colX) Leave(in ast.Node) (ast.Node, bool) {
	return in, true
}
```

Finally, wrap `colX` in a simple function:

```go
func extract(rootNode *ast.StmtNode) []string {
	v := &colX{}
	(*rootNode).Accept(v)
	return v.colNames
}
```

and slightly modifiy main function:

```go
func main() {
	if len(os.Args) != 2 {
		fmt.Println("usage: colx 'SQL statement'")
		return
	}
	sql := os.Args[1]
	astNode, err := parse(sql)
	if err != nil {
		fmt.Printf("parse error: %v\n", err.Error())
		return
	}
	fmt.Printf("%v\n", extract(astNode))
}
```

Test it:

```bash
go build && ./colx 'select a, b from t'
```

```
[a b]
```

You can also try with different SQL as an input, i.e.

```console
$ ./colx 'SELECT a, b FROM t GROUP BY (a, b) HAVING a > c ORDER BY b'

> [a b a b a c b]

Reviewer 2: should mention it will be duplucate?

# Quickstart

This parser is highly compatible with MySQL syntax. You can use it as a library, parse a text SQL into an AST tree, and traverse the AST nodes.

In this example, you will build a project, which can extract all the column names from a text SQL.

## Prerequisites

- [Golang](https://golang.org/dl/) version 1.13 or above. You can follow the instructions in the official [installation page](https://golang.org/doc/install) (check it by `go version`)

## Create a Project

```bash
mkdir colx && cd colx
go mod init colx && touch main.go
```

## Import Dependencies

First of all, you need to use `go get` to fetch the dependencies:

```bash
go get -v github.com/pingcap/parser@v4.0.0-rc.1
```

> **NOTE**
>
> You may want to use advanced API on expressions (a kind of AST node), such as numbers, string literals, booleans, nulls, etc. It is strongly recommended to use the `types` package in TiDB repo with the following command:
>
> ```bash
> go get -v github.com/pingcap/tidb/types/parser_driver@v4.0.0-rc.1
> ```

> > and import it in the directive of your golang source code:

Author: ```suggestion
> and import it in your golang source code:
```

> > and import it in the directive of your golang source code:

Author: It means declarations like 
```go
import github.com/pingcap/tidb/types/parser_driver
```

I think "and import it in ~the directive of~ your golang source code:" is clear enough.

> ```go
> import _ "github.com/pingcap/tidb/types/parser_driver"
> ```

Your directory should contain the following three files:
```
.
├── go.mod
├── go.sum
└── main.go
```

> Now, open `main.go` with your favouriate editor, and start coding!

Reviewer 3: ```suggestion
Now, open `main.go` with your favorite editor, and start coding!
```

	_ "github.com/pingcap/parser/test_driver"
)

func parse(sql string) (*ast.StmtNode, error) {
	p := parser.New()

	stmtNodes, _, err := p.Parse(sql, "", "")
	if err != nil {
		return nil, err
	}

	return &stmtNodes[0], nil
}

func main() {
	astNode, err := parse("SELECT a, b FROM t")
	if err != nil {
		fmt.Printf("parse error: %v\n", err.Error())
		return
	}
	fmt.Printf("%v\n", *astNode)
}
```

Test the parser by running the following command:

```bash
go run main.go
```

If the parser runs properly, you should get a result like this:

```
&{{{{SELECT a, b FROM t}}} {[]} 0xc0000a1980 false 0xc00000e7a0 <nil> 0xc0000a19b0 <nil> <nil> [] <nil> <nil> none [] false false 0 <nil>}
```

> **NOTE**
>
> Here are a few things you might want to know:
> - To use a parser, a `parser_driver` is required. It decides how to parse the basic data types in SQL.
>
>   You can use [`github.com/pingcap/parser/test_driver`](https://pkg.go.dev/github.com/pingcap/parser/test_driver) as the `parser_driver` for test. Again, if you need advanced features, please use the `parser_driver` in TiDB (run `go get -v github.com/pingcap/tidb/types/parser_driver@v4.0.0-rc.1` and import it).

> > - The instantiated parser object is not goroutine safe. it is better to keep it in a single goroutine.

Reviewer 3: ```suggestion
> - The instantiated parser object is not goroutine safe. It is better to keep it in a single goroutine.
```

> > - The instantiated parser object is not lightweight, it is better to reuse it if possible.

Reviewer 3: ```suggestion
> - The instantiated parser object is not lightweight. It is better to reuse it if possible.
```

Parser implements the interface [`ast.Node`](https://pkg.go.dev/github.com/pingcap/parser/ast?tab=doc#Node) for each kind of AST node, such as SelectStmt, TableName, ColumnName. [`ast.Node`](https://pkg.go.dev/github.com/pingcap/parser/ast?tab=doc#Node) provides a method `Accept(v Visitor) (node Node, ok bool)` to allow any struct that has implemented [`ast.Visitor`](https://pkg.go.dev/github.com/pingcap/parser/ast?tab=doc#Visitor) to traverse itself.

[`ast.Visitor`](https://pkg.go.dev/github.com/pingcap/parser/ast?tab=doc#Visitor) is defined as follows:
```go
type Visitor interface {
	Enter(n Node) (node Node, skipChildren bool)
	Leave(n Node) (node Node, ok bool)
}
```

Now you can define your own visitor, `colX`(columnExtractor):

```go
type colX struct{
	colNames []string
}

func (v *colX) Enter(in ast.Node) (ast.Node, bool) {
	if name, ok := in.(*ast.ColumnName); ok {
		v.colNames = append(v.colNames, name.Name.O)
	}
	return in, false
}

func (v *colX) Leave(in ast.Node) (ast.Node, bool) {
	return in, true
}
```

Finally, wrap `colX` in a simple function:

```go
func extract(rootNode *ast.StmtNode) []string {
	v := &colX{}
	(*rootNode).Accept(v)
	return v.colNames
}
```

> And slightly modifiy the main function:

Reviewer 3: ```suggestion
And slightly modify the main function:
```

```go
func main() {
	if len(os.Args) != 2 {
		fmt.Println("usage: colx 'SQL statement'")
		return
	}
	sql := os.Args[1]
	astNode, err := parse(sql)
	if err != nil {
		fmt.Printf("parse error: %v\n", err.Error())
		return
	}
	fmt.Printf("%v\n", extract(astNode))
}
```

> Test it:

Author: ```suggestion
Test your program:
```
